### ISSUE TITLE
Subset assignment into a QualityScaledStringSet does not work

### ISSUE THREAD

**Issue by LTLA** [Bioconductor/Biostrings#2]
I would have expected this to work;

```r
library(Biostrings)
stuff <- QualityScaledDNAStringSet(DNAStringSet(c("AAA", "TTT", "CCC", "GGG")),
                                   PhredQuality(c("(((", "!!!", "###", "...")))
stuff[1] <- stuff[2]
```

... except that I get the following complaint:

```
Error in validObject(.Object) : 
  invalid class “QualityScaledDNAStringSet” object: invalid object for slot "quality" in class "QualityScaledDNAStringSet": got class "S4", should be or extend class "XStringQuality"
```

There are no problems with the equivalent `DNAStringSet` operation:

```r
whee <- as(stuff, "DNAStringSet")
whee[1] <- whee[2] # Works as expected.
```
 
Session info:

```
R Under development (unstable) (2017-10-27 r73632)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 14.04.5 LTS

Matrix products: default
BLAS: /home/cri.camres.org/lun01/Software/R/trunk/lib/libRblas.so
LAPACK: /home/cri.camres.org/lun01/Software/R/trunk/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_GB.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_GB.UTF-8        LC_COLLATE=en_GB.UTF-8    
 [5] LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_GB.UTF-8   
 [7] LC_PAPER=en_GB.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] stringr_1.2.0              ShortRead_1.37.0          
 [3] GenomicAlignments_1.15.2   SummarizedExperiment_1.9.5
 [5] DelayedArray_0.5.14        matrixStats_0.52.2        
 [7] Biobase_2.39.1             Rsamtools_1.31.1          
 [9] BiocParallel_1.13.1        sarlacc_1.0.0             
[11] Biostrings_2.47.1          XVector_0.19.1            
[13] GenomicRanges_1.31.3       GenomeInfoDb_1.15.1       
[15] IRanges_2.13.4             S4Vectors_0.17.15         
[17] BiocGenerics_0.25.1       

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.14           RColorBrewer_1.1-2     compiler_3.5.0        
 [4] pillar_1.0.1           prettyunits_1.0.2      GenomicFeatures_1.31.1
 [7] bitops_1.0-6           tools_3.5.0            zlibbioc_1.25.0       
[10] progress_1.1.2         biomaRt_2.35.5         digest_0.6.13         
[13] bit_1.1-12             RSQLite_2.0            memoise_1.1.0         
[16] tibble_1.4.1           lattice_0.20-35        pkgconfig_2.0.1       
[19] rlang_0.1.6            igraph_1.1.2           Matrix_1.2-12         
[22] DBI_0.7                GenomeInfoDbData_1.1.0 hwriter_1.3.2         
[25] rtracklayer_1.39.4     httr_1.3.1             ade4_1.7-10           
[28] bit64_0.9-7            grid_3.5.0             R6_2.2.2              
[31] AnnotationDbi_1.41.4   XML_3.98-1.9           RMySQL_0.10.13        
[34] latticeExtra_0.6-28    seqinr_3.4-5           magrittr_1.5          
[37] blob_1.1.0             MASS_7.3-48            assertthat_0.2.0      
[40] muscle_3.21.0          stringi_1.1.6          RCurl_1.95-4.9        
```

**Comment by hpages** [comment]
and will look at this too...

**Comment by LTLA** [comment]
The actual problem occurs within `XVector::unlist_list_of_XVectorList`, but I didn't know how to fix that, so I just created a new `[<-` method for `QualityScaledXStringSet` objects instead.

**Comment by hpages** [comment]
Hi @LTLA,
Just to let you know that I'm working on this but it has taken me to a longer road than anticipated. The basic problem is that `c()` is broken on QualityScaledStringSet objects. Turns out that I had on my radar to refactor how `"c"` methods are implemented for Vector derivatives for quite a while but was always postponing. The idea is to simplify the whole thing by taking advantage of `parallelSlotNames()`. With this refactoring, `c()` will work out-of-the-box on QualityScaledStringSet objects. So instead of implementing a "c" method for QualityScaledStringSet objects that I would need to get rid of later, I started this refactoring 2 days ago. I already made changes to S4Vectors, IRanges, and XVector but still need to take care of Biostrings, GenomicRanges, GenomicAlignments and probably more (including InteractionSet and GenomicTuples, at least). In the particular case of QualityScaledStringSet, what makes things a little bit more complicated is that the `quality` slot is currently not required to be parallel to the object (it's allowed to be of length 1). This was a design decision made a long time ago when this class was added to Biostrings, as an attempt to save some memory I guess. IMO this introduces a lot of complexity in the internals and saves very little memory (the XStringQuality object can be `rep`'ed very cheaply at construction time before being stored in the `quality` slot) so is not worth it. So I'm also in the process of changing this to make the `quality` slot a parallel slot. This implies some significant changes to the QualityScaledStringSet code. I'll let you know when I'm done. Thanks for your patience.

**Comment by LTLA** [comment]
Woah, sounds intense. Thanks for the update.

**Comment by hpages** [comment]
This should be fixed in Biostrings 2.47.2.

**Comment by LTLA** [comment]
Checked it and it works perfectly, thanks.
